### ISSUE TITLE
Problem with pickle thread.lock

### ISSUE THREAD

**Issue by kivio** [GoogleCloudPlatform/PerfKitBenchmarker#424]
When i start benchmark by OpenStack driver i get:
```Can't pickle <type 'thread.lock'>: it's not found as thread.lock```

This problem due from python-novaclient and objects with connection to API.
My proposition is to use dill instead of pickle for pickling.
Dill support ThreadLock's in proper way, benchmark with dill work's fine. I was testing it on my setup.

What do you thing about that?

**Comment by cmccoy** [comment]
Dill looks like a nice option to me.
We've been handling saving/restoring locks manually (e.g., [here](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/linux_virtual_machine.py#L94-L106)), which is a bit crude.

**Comment by kivio** [comment]
Another option and i think more secure (because override just Lock pickling) is use copy_reg and functions for pickle and unpickle Lock.

```
import copy_reg
import thread
import threading

def pickle_lock(lock):
    return threading.Lock, (lock.locked(),)

def unpickle_lock(locked, *args):
    lock = threading.Lock()
    if locked:
        if not lock.acquire(False):
            raise pickle.UnpicklingError("Cannot acquire lock")

copy_reg.pickle(thread.LockType, pickle_lock, unpickle_lock)
```

**Comment by sirtux** [comment]
Hi, I also have seen this problem - is there a proposed fix? The fork from kivio doesn't support openstack as it seems..

Thanks,
Tom

**Comment by cmccoy** [comment]
Closed by #425. @sirtux: if you're still having problems with the current master branch, could you open a separate issue with details?
